ch_pipeline/hfb/containers.py

```diff
@@ -51,7 +54,50 @@ def nsample(self) -> memh5.MemDataset:
             raise KeyError("Dataset 'nsample' not initialised.")
 
 
-class HFBData(RawContainer, FreqContainer, HFBContainer):
+class HFBBeamContainer(HFBContainer):
+    """A pipeline container for HFB data with a beam axis.
+
+    This works like a normal :class:`HFBContainer` container, but already has a beam
+    axis defined, and specific properties for dealing with beams.
+    """
+
+    _axes = ("beam",)
+
+    @property
+    def beam(self):
+        """The beam indices associated with each entry of the beam axis."""
+        return self.index_map["beam"]
+
+    @property
+    def beam_ew(self):
+        """The unique EW-beam indices (i.e., from 0 to 3) in the beam axis."""
+        return self.get_beam_ew()
+
+    @property
+    def beam_ns(self):
+        """The unique NS-beam indices (i.e., from 0 to 256) in the beam axis."""
+        return self.get_beam_ns()
+
+    def get_beam_ew(self, return_inverse=False):
+        """Find the unique EW-beam indices (i.e., from 0 to 3) in the beam axis.
+
+        If called with `return_inverse=True`, return a tuple of (a) unique EW-beam
+        indices and (b) an array mapping each sample in the beam axis to an index
+        in the output array of unique EW-beam indices.
+        """
+        return np.unique(self.beam // 256, return_inverse=return_inverse)
+
+    def get_beam_ns(self, return_inverse=False):
+        """Find the unique NS-beam indices (i.e., from 0 to 256) in the beam axis.
+
+        If called with `return_inverse=True`, return a tuple of (a) unique NS-beam
+        indices and (b) an array mapping each sample in the beam axis to an index
+        in the output array of unique NS-beam indices.
+        """
+        return np.unique(self.beam % 256, return_inverse=return_inverse)
```

Comment:
I suggest having only the properties and remove the get methods to tighten up the API. Also, you should cache the properties to avoid regenerating at every call. There is a `cached_property` decorator for that in `functools` that we use elsewhere in the pipeline.

Reply:
These changes are now implemented in PR #222